File: src/infrastructure/persistence/filesystem_file_writer.py
"""Filesystem implementation of FileWriterPort.

Writes files to disk with automatic parent directory creation.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from pathlib import Path


class FilesystemFileWriter:
    """Writes files to the local filesystem.

    Implements FileWriterPort by writing content to the given path,
    creating parent directories as needed.
    """

    def write_file(self, path: Path, content: str) -> None:
        """Write content to a file at the given path.

        Creates parent directories if they don't exist. Overwrites
        existing files. Lets OSError propagate on permission failures.

        Args:
            path: Target file path.
            content: File content to write.
        """
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(content, encoding="utf-8")

File: internal/shared/infrastructure/persistence/filesystem_file_writer.go
// Package persistence provides filesystem-based infrastructure adapters
// for the shared bounded context.
package persistence

import (
	"context"
	"fmt"
	"os"
	"path/filepath"

	sharedapp "github.com/alto-cli/alto/internal/shared/application"
)

// Compile-time interface satisfaction check.
var _ sharedapp.FileWriter = (*FilesystemFileWriter)(nil)

// FilesystemFileWriter writes files to the local filesystem.
// Implements FileWriter by writing content to the given path,
// creating parent directories as needed.
type FilesystemFileWriter struct{}

// NewFilesystemFileWriter creates a new FilesystemFileWriter.
func NewFilesystemFileWriter() *FilesystemFileWriter {
	return &FilesystemFileWriter{}
}

// WriteFile writes content to a file at the given path.
// Creates parent directories if they don't exist. Overwrites existing files.
func (w *FilesystemFileWriter) WriteFile(_ context.Context, path string, content string) error {
	dir := filepath.Dir(path)
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return fmt.Errorf("creating directory %s: %w", dir, err)
	}
	if err := os.WriteFile(path, []byte(content), 0o644); err != nil {
		return fmt.Errorf("writing file %s: %w", path, err)
	}
	return nil
}
